### ISSUE TITLE
Plugin IO should use timeouts

### ISSUE THREAD

**Issue by bmr-cymru** [sosreport/sos#180]
This is going to be tough to implement but it would be helpful to have timeouts applied to the IO carried out when collecting files for a plug-in. Right now we use a mix of simple IO (e.g. in substitution code paths) and shutils. None of these APIs have built-in timeouts afaik.

The reason for this is that we may hit files on the system that hang when read; either because of sos bugs (see commit dc9d212) or because the system is in some broken state.

We can't make this perfect but currently any problem like this totally breaks sos - the command just hangs indefinitely.

We've improved behaviour for hanging commands by running under the timeout wrapper. It would be good to do something similar for file IO.

One option would be to treat everything (inc. file copies) as a command.. Then they can all use the timeout wrapper trick. It's a bit of a hack although it's also how sos handled file manipulation for a long time (until 2.0 we were shelling out to cp -p..)

**Comment by TurboTurtle** [comment]
I'm thinking that the whole-plugin timeouts provided by 04df944 should provide what we need here? If a file copy hangs, it will only hang one thread up to the plugin's `timeout` (default 300s) and then the executor kills the thread. We end up not proceeding with any subsequent collections (either file or command) the plugin would have done, but it still allows sos to continue.

**Comment by bmr-cymru** [comment]
Has this been tested with a "true" kernel side hang? I.e. are we confident that the thread kill will be honoured in the even that a thread is blocked on IO and in D-state?

**Comment by hegdevasant** [comment]
@TurboTurtle We recently had FS issue and then tried to collect sosreport and it got hung. It turned out that command was blocked in IO and didn't respond to kill signal. As @bmr-cymru suggested we passed  `-k` option to `timeout` command. It sent SIGKILL to the process group created by `timeout`. But sosreport was blocked in Popen.read() call and hence didn't respond to SIGCHLD.. So  sosreport didn't came out.

I didn't try to introduce separate Thread .. but there is a concern about using Popen.wait instead of Popen.poll (See below patch). Let me know if you have better solution. I can help you to test the patch.


-Vasant




From: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
Date: Tue, 24 Jul 2018 16:20:04 +0530
Subject: [PATCH] sosreport: Fix hang issue

sosreport collects number of command output. It uses timeout command
to terminate the commands which are not completing within certain time.
But if command is blocked in kernel then it will not respond to SIGTERM
from `timeout` command. Hence sosreport hangs.

To fix above situation, this patch adds `-k <seconds>` option to timeout
command so that it will eventually send SIGKILL to process group created
by timeout command. Now `timeout` command dies and `init` process
becomes parent of actual command. Finally `timeout` command sends signal
to sosreport.

But sosreport uses Popen.read() call to read data from child process. By
default read() is blocking call [1]. And actual PIPE is not yet closed.
Hence sosreport will not respond to the signal and it hangs.

To overcome the above situation this patch creates separate thread to
read data from child process. So that parent thread handles signals.
This replaces poll() with wait() function so that we wait until child
thread completes reading data.

Finally adjusts return code so that caller can log apropriate warning
message.

[1] http://eyalarubas.com/python-subproc-nonblock.html

Signed-off-by: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
---
 sos/utilities.py | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/sos/utilities.py b/sos/utilities.py
index d112e15a..6d5f6690 100644
--- a/sos/utilities.py
+++ b/sos/utilities.py
@@ -18,6 +18,7 @@ import errno
 import shlex
 import glob
 import threading
+from threading import Thread
 
 from contextlib import closing
 from collections import deque
@@ -133,7 +134,7 @@ def sos_get_command_output(command, timeout=300, stderr=False,
                 cmd_env.pop(key, None)
     # use /usr/bin/timeout to implement a timeout
     if timeout and is_executable("timeout"):
-        command = "timeout %ds %s" % (timeout, command)
+        command = "timeout -k 5 %ds %s" % (timeout, command)
 
     # shlex.split() reacts badly to unicode on older python runtimes.
     if not six.PY3:
@@ -154,8 +155,8 @@ def sos_get_command_output(command, timeout=300, stderr=False,
                   preexec_fn=_child_prep_fn)
 
         reader = AsyncReader(p.stdout, sizelimit, binary)
+        p.wait()
         stdout = reader.get_contents()
-        p.poll()
 
     except OSError as e:
         if e.errno == errno.ENOENT:
@@ -166,6 +167,10 @@ def sos_get_command_output(command, timeout=300, stderr=False,
     if p.returncode == 126 or p.returncode == 127:
         stdout = six.binary_type(b"")
 
+    # Convert SIGKILL return code timeout
+    if p.returncode == -9:
+        p.returncode = 124
+
     return {
         'status': p.returncode,
         'output': stdout
@@ -214,8 +219,9 @@ class AsyncReader(threading.Thread):
             sizelimit = sizelimit * 1048576  # convert to bytes
             slots = int(sizelimit / self.chunksize)
         self.deque = deque(maxlen=slots)
-        self.start()
-        self.join()
+        self._t = Thread(target = self.run)
+        self._t.daemon = True
+        self._t.start()
 
     def run(self):
         '''Reads from the channel (pipe) that is the output pipe for a
-- 
2.14.3

**Comment by bmr-cymru** [comment]
Signals should never be delivered to the worker threads anyway (unless something has gone horribly wrong), so I think it's unlikely that the explanation that we're blocked in `read()` and hence ignoring `SIGCHLD` is not entirely correct.

**Comment by hegdevasant** [comment]
@bmr-cymru I tried upstream master on my system. By default it created 4 worker threads. But even with `timeout -k` option it hung ...as  commands from multiple plugin blocked in IO read. I did observe 4 `timeout` process in Zombie state. That made me to think that signal was delivered to thread which started Popen and it was blocked in `read` call.
I can't think of any other reason for upstream sosreport to hang. Am I missing something?


-Vasant
